radicle-node/src/service.rs

```diff
@@ -1110,8 +1113,68 @@ where
                 }
                 debug!(target: "service", "Fetch accepted for {rid} from {remote}..");
 
+                let Ok(t) = self.tracking.repo_policy(&rid) else {
+                    if let Some(resp) = self.fetch_reqs.get(&rid) {
+                        resp.send(FetchResult::Failed {
+                            reason: format!("Failed to look tracking policy for {rid}")
+                        }).ok();
+                    }
+                    return Ok(())
+                };
+                let ns = match t.policy {
+                    Policy::Block => {
+                        error!(target: "service", "Attempted to fetch blocked repo {rid}");
+
+                        if let Some(resp) = self.fetch_reqs.get(&rid) {
+                            resp.send(FetchResult::Failed {
+                                reason: String::from("Attempted to fetch blocked repo"),
+                            })
+                            .ok();
+                        }
+                        return Ok(());
+                    }
+                    Policy::Track => match t.scope {
+                        Scope::All => Namespaces::All,
+                        Scope::Trusted => {
+                            let mut trusted: Vec<_> = self
+                                .tracking
+                                .node_entries()
+                                .unwrap()
+                                .filter_map(|node| {
+                                    (node.policy == Policy::Track).then_some(node.id)
+                                })
+                                .collect();
+
+                            let ns = if let Ok(repo) = self.storage.repository(rid) {
+                                let delegates = repo.delegates().unwrap().map(PublicKey::from);
+                                trusted.extend(delegates);
+
+                                NonEmpty::from_vec(trusted).map(Namespaces::Many)
+                            } else {
+                                Some(Namespaces::All)
+                            };
+
+                            if let Some(ns) = ns {
+                                ns
+                            } else {
+                                warn!(target: "service", "Attempted to fetch repo {rid} with no trusted peers");
+
+                                if let Some(resp) = self.fetch_reqs.get(&rid) {
+                                    resp.send(FetchResult::Failed {
+                                        reason: String::from(
+                                            "Attempted to fetch with no trusted peers",
+                                        ),
+                                    })
+                                    .ok();
+                                }
+                                return Ok(());
+                            }
+                        }
+                    },
+                };
```

Comment:
This block contains a few `unwraps` that we should document with `// SAFETY`, turn to expects, or convert to errors

Reply:
I extracted this big ass block out into a function that returns the `Namespaces` or a `FetchResult` (really the failure cases) and use that function instead.